mllib/src/main/scala/org/apache/spark/ml/tree/impl/SplitUtils.scala

```diff
@@ -0,0 +1,215 @@
+        if (parentCalc.isEmpty) {
+          parentCalc = Some(gainAndImpurityStats.impurityCalculator)
+        }
+        (splitIndex, gainAndImpurityStats)
+      }.maxBy(_._2.gain)
+    (featureSplits(bestFeatureSplitIndex), bestFeatureGainStats)
+
+  }
+
+  /**
+   * Choose splitting rule: feature value <= threshold
+   *
+   * @return  (best split, statistics for split)  If the best split actually puts all instances
+   *          in one leaf node, then it will be set to None.  If no valid split was found, the
+   *          returned ImpurityStats instance will be invalid (have member valid = false)
+   */
+  private[impl] def chooseContinuousSplit(
+      binAggregates: DTStatsAggregator,
+      featureIndex: Int,
+      featureIndexIdx: Int,
+      featureSplits: Array[Split],
+      parentCalculator: Option[ImpurityCalculator] = None): (Split, ImpurityStats) = {
+    // For a continuous feature, bins are already sorted for splitting
+    // Number of "categories" = number of bins
+    val sortedCategories = Range(0, binAggregates.metadata.numBins(featureIndex)).toList
+    // Get & return best split info
+    val (bestFeatureSplitIndex, bestFeatureGainStats) = orderedSplitHelper(binAggregates,
+      featureIndex, featureIndexIdx, sortedCategories, parentCalculator)
+    (featureSplits(bestFeatureSplitIndex), bestFeatureGainStats)
+  }
+
+  /**
+   * Computes the index of the best split for an ordered feature.
+   * @param parentCalculator Optional: ImpurityCalculator containing impurity stats for current node
+   */
+  private def orderedSplitHelper(
+      binAggregates: DTStatsAggregator,
+      featureIndex: Int,
+      featureIndexIdx: Int,
+      categoriesSortedByCentroid: List[Int],
+      parentCalculator: Option[ImpurityCalculator]): (Int, ImpurityStats) = {
+    // Cumulative sum (scanLeft) of bin statistics.
+    // Afterwards, binAggregates for a bin is the sum of aggregates for
+    // that bin + all preceding bins.
+    assert(!binAggregates.metadata.isUnordered(featureIndex))
+    val numSplits = binAggregates.metadata.numSplits(featureIndex)
+    val nodeFeatureOffset = binAggregates.getFeatureOffset(featureIndexIdx)
+    var splitIndex = 0
+    while (splitIndex < numSplits) {
+      val currentCategory = categoriesSortedByCentroid(splitIndex)
+      val nextCategory = categoriesSortedByCentroid(splitIndex + 1)
+      binAggregates.mergeForFeature(nodeFeatureOffset, nextCategory, currentCategory)
+      splitIndex += 1
+    }
+    // lastCategory = index of bin with total aggregates for this (node, feature)
+    val lastCategory = categoriesSortedByCentroid.last
+
+    // Find best split.
+    var parentCalc = parentCalculator
+    Range(0, numSplits).map { splitIndex =>
+      val featureValue = categoriesSortedByCentroid(splitIndex)
+      val leftChildStats =
+        binAggregates.getImpurityCalculator(nodeFeatureOffset, featureValue)
+      val rightChildStats =
```

Comment:
Exactly, it's the parentCalc minus the left child stats. Since `ImpurityCalculator.subtract()` updates the impurity calculator in place, we call `binAggregates.getParentImpurityCalculator()` to get a copy of the parent impurity calculator, then subtract the left child stats.